heat/core/factories.py

```diff
@@ -309,6 +320,20 @@ def array(
         if obj.dtype != torch_dtype:
             obj = obj.type(torch_dtype)
 
+    # infer device from obj if not explicitly given
+    if device is None:
+        device = devices.sanitize_device(obj.device.type)
+
+    # change device if it do not match
+    assert str(obj.device) == device.torch_device
+
+    if str(obj.device) != device.torch_device:
+        warnings.warn(
+            "Array 'obj' is not on device '{}'. It will be copied to it.".format(device),
+            UserWarning,
+        )
```

Comment:
Should I keep the assertion or the warning here?

Reply:
I deleted the assertion as there are some cases where it is allowed. However the unittest results must be reviewed for possible wrong devices.